### ISSUE TITLE
enum members added by `enum.global_enum` to module space is not detected by pyright.

### ISSUE THREAD

**Issue by Achxy** [microsoft/pyright#4351]
### Discussed in https://github.com/microsoft/pyright/discussions/4239

<div type='discussions-op-text'>

<sup>Originally posted by **Achxy** November 27, 2022</sup>
If enum members were exported to module global namespace by [`enum.global_enum`](https://docs.python.org/3/library/enum.html#enum.global_enum) they are not reflected by pyright and are considered as not defined.
Example:
```py
from enum import Enum, auto, global_enum


@global_enum
class Foo(Enum):
    BAR = auto()
    QUX = auto()


print(QUX)  # "QUX" is not defined
# However it is defined during runtime
```
</div>

---------------

Since the values of an enum class cannot be patched once defined and that enum classes cannot be further subclassed should make this feature at least possible.
```py
class Foo(Enum):
    A = 1

Foo.A.value = 2 # Fails during runtime
```
and,
```py
class Foo(Enum):
    A = 1

class Bar(Foo): # Fails during typecheck and during runtime
    B = 2
```
There are instances of this feature [being used in the standard library - cpython](https://github.com/python/cpython/blob/75a6fadf369315b27e12f670e6295cf2c2cf7d7e/Lib/re/__init__.py#L141-L156)
and implementation of this feature could aid in eloquence by [preventing such annotations - typeshed](https://github.com/python/typeshed/blob/ea7bbbdad7abf05a3bd3feb25eef93b53165e2d9/stdlib/re.pyi#L165-L202)
```py
A = RegexFlag.A
ASCII = RegexFlag.ASCII
DEBUG = RegexFlag.DEBUG
I = RegexFlag.I
IGNORECASE = RegexFlag.IGNORECASE
L = RegexFlag.L
LOCALE = RegexFlag.LOCALE
M = RegexFlag.M
MULTILINE = RegexFlag.MULTILINE
S = RegexFlag.S
DOTALL = RegexFlag.DOTALL
X = RegexFlag.X
VERBOSE = RegexFlag.VERBOSE
U = RegexFlag.U
UNICODE = RegexFlag.UNICODE
T = RegexFlag.T
TEMPLATE = RegexFlag.TEMPLATE
```
Making enum contents public some use-cases like observed in `re` and some popular libraries. So this may be a great feature to be implemented to support such cases.

**Comment by erictraut** [comment]
The `@enum.global_enum` decorator is a very problematic feature. I wish it would have been discussed publicly before it was added so we could have tried to prevent its addition. It breaks all existing scoping rules in Python by implicitly modifying a different scope's namespace. Such a feature breaks assumptions that are core to pyright's architecture. We therefore don't have plans to support it. I strongly recommend against using it.

We may consider adding a warning for `@global_enum` to indicate that it is not supported in static type checking and should not be used.
